mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -68,16 +69,70 @@ void ZneLowering::rewrite(mitigation::ZneOp op, PatternRewriter &rewriter) const
+                        }
+                        else {
+                            resultExtracted = resultValue;
+                        }
+                        Value resultInserted = builder.create<tensor::InsertOp>(
+                            loc, resultExtracted, iterArgs.front(), index);
+                        builder.create<scf::YieldOp>(loc, resultInserted);
+                    }
+                    else {
+                        // Multiple measurements
+                        ValueRange resultValuesMulti = callOp.getResults();
+                        SmallVector<Value> vectorResultsMulti;
+                        // Create a tensor
+                        for (Value resultValue : resultValuesMulti) {
+                            Value resultExtracted;
+                            if (isa<RankedTensorType>(resultValue.getType())) {
+                                resultExtracted =
+                                    builder.create<tensor::ExtractOp>(loc, resultValue);
+                            }
+                            else {
+                                resultExtracted = resultValue;
+                            }
+                            vectorResultsMulti.push_back(resultExtracted);
+                        }
+                        SmallVector<int64_t> resShape = {numResults};
+                        Type type =
+                            RankedTensorType::get(resShape, vectorResultsMulti[0].getType());
+                        auto tensorResults =
+                            builder.create<tensor::FromElementsOp>(loc, type, vectorResultsMulti);
+                        Value sizeResultsValue =
+                            rewriter.create<index::ConstantOp>(loc, numResults);
+                        Value resultValuesFor =
+                            rewriter
+                                .create<scf::ForOp>(
+                                    loc, c0, sizeResultsValue, c1,
+                                    /*iterArgsInit=*/iterArgs.front(),
+                                    [&](OpBuilder &builder, Location loc, Value j,
+                                        ValueRange iterArgsIn) {
+                                        Value resultExtracted = builder.create<tensor::ExtractOp>(
+                                            loc, tensorResults, j);
+                                        SmallVector<Value> indices = {i, j};
+                                        Value resultInserted = builder.create<tensor::InsertOp>(
+                                            loc, resultExtracted, iterArgsIn.front(), indices);
+
+                                        builder.create<scf::YieldOp>(loc, resultInserted);
+                                    })
+                                .getResult(0);
+                        builder.create<scf::YieldOp>(loc, resultValuesFor);
+                    }
```

Comment:
Nit pick:

I would like this to be separated into different functions. It is a bit hard to read. Also, creating a for loop for two elements is a bit arbitrary. I am happy keeping the for-loop, but at some point we should investigate when to generate a for loop vs just unrolling the loop.

Reply:
I have merged the single and multiple measurement case to make it cleaner. Let me know what you think.

Potentially the scale factors is big (first loop), return/measurements (second loop) is potentially smaller but it can be anything.